### PR TITLE
allow persist data volume when delete the VM

### PR DESCRIPTION
**Pull request by aiwantaozi**
Issue: https://github.com/rancher/harvester/issues/69

### CODE REVIEW

File: pkg/api/vm/store.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+func (s *vmStore) Delete(request *types.APIRequest, schema *types.APISchema, id string) (types.APIObject, error) {
+	deleteDisk := request.Query.Get("deleteDisk")
+	if deleteDisk == "false" {
+		if err := s.updateDataVolumeOwnerRef(request.Namespace, request.Name); err != nil {
+			return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to update dataVolume's ownerReferences, %v", err))
+		}
+	}
+
+	return s.Store.Delete(request, request.Schema, id)
```
Comment (gitlawr) [inline pkg/api/vm/store.go]: This assumes that DVs are defined in VM spec. It does not consider the case when DV is created independently and the users want to remove them on VM deletion.

File: pkg/api/vm/store.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+			if slice.ContainsString(deleteDisks, vol.Name) {
+				switch {
+				case vol.PersistentVolumeClaim != nil:
+					deletePvc = append(deletePvc, vol.PersistentVolumeClaim.ClaimName)
+				case vol.Secret != nil:
+					deleteSecret = append(deleteSecret, vol.Secret.SecretName)
+				case vol.ConfigMap != nil:
+					deleteConfigmap = append(deleteConfigmap, vol.ConfigMap.Name)
+				case vol.ServiceAccount != nil:
+					deleteServiceAccount = append(deleteServiceAccount, vol.ServiceAccount.ServiceAccountName)
```
Comment (gitlawr) [inline pkg/api/vm/store.go]: How do we tell users what we are gonna do when they select these disks? Given that these are not volumes

Comment (aiwantaozi) [reply pkg/api/vm/store.go]: Will support later if we need, remove pvc secret, configmap, serviceAccount manage now.

File: pkg/api/vm/store.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	apiObj, err := s.Store.Delete(request, request.Schema, id)
+	if err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to delete vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
+
+	if err := s.deleteDataVolumes(vm.Namespace, deleteDataVolume); err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to delete dataVolume for vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
```
Comment (guangbochen) [inline pkg/api/vm/store.go]: should delete DV first and remove the VM is succeeded.

File: pkg/api/vm/store.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	return apiObj, nil
+}
+
+func (s *vmStore) updateDataVolumeOwnerRef(vmNamespace, vmName string, keepDataVolumes []string) error {
```
Comment (guangbochen) [inline pkg/api/vm/store.go]: prefer to call it `removeVMDataVolumeOwnerRef` since this is what the method does.

File: pkg/api/vm/store.go (hunk 3 above)
Comment (aiwantaozi) [reply pkg/api/vm/store.go]: vm depends on the dv, prefer to delete vm then dv to prevent vm go into invalid dv error.

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**